test/lib/jdk/test/lib/hexdump/ASN1Formatter.java

```diff
@@ -0,0 +1,611 @@
+            if (len > 0x80) {
+                // multi-byte encoded length
+                int nbytes = len & 0x7f;
+                if (nbytes > 4) {
+                    out.append("***** Tag: " + tagName(tag) +
+                            ", Range of length error: " + len + "bytes");
+                    return available;       // return the unread length
+                }
+                len = 0;
+                for (; nbytes > 0; nbytes--) {
+                    int inc = in.readByte() & 0xff;
+                    len = (len << 8) | (0xff & inc);
+                    available -= nbytes;
+                }
+            } else if (len == 0x80) {
+                // Tag with Indefinite-length; note the tag and continue parsing tags.
+                out.append(prefix).append(tagName(tag)).append(": INDEFINITE-LENGTH CONTENT\n");
+                continue;
+            }
+            if (available < 0 && origAvailable < 0) {
+                // started out unknown; set available to the length of this tagged value
+                available = len;
+            }
+            out.append(prefix);     // start with indent
+            switch (tag) {
+                case TAG_EndOfContent:    // End-of-contents octets; len == 0
+                    out.append("END-OF-CONTENT ");
+                    // end of indefinite-length constructed, return any remaining
+                    return available;
+                case TAG_Integer:
+                case TAG_Enumerated:
+                    switch (len) {
+                        case 1:
+                            out.append(String.format("BYTE %d, ", in.readByte()));
+                            available -= 1;
+                            break;
+                        case 2:
+                            out.append(String.format("SHORT %d, ", in.readShort()));
+                            available -= 2;
+                            break;
+                        case 4:
+                            out.append(String.format("INTEGER %d, ", in.readInt()));
+                            available -= 4;
+                            break;
+                        case 8:
+                            out.append(String.format("LONG %d, ", in.readLong()));
+                            available -= 8;
+                            break;
+                        default:
+                            formatBytes(in, out, len);
+                            available -= len;
+                            break;
+                    }
+                    break;
+                case TAG_ObjectId:
+                    byte[] oid = new byte[len];
+                    int l1 = in.read(oid);
+                    available -= l1;
+                    String s = oidName(oid);
+                    out.append(tagName(tag) + " [" + len + "] ");
+                    out.append(s);
+                    out.append(' ');
+                    break;
+
+                case TAG_OctetString:
```

Comment:
Octet Strings can may contain printable strings and it is useful to print them.
A heuristic is used to determine if they are printable and the length is limited to avoid spewing garbage.